pkg/proxy/iptables/proxier.go

```diff
@@ -964,29 +1093,97 @@ func (proxier *Proxier) syncProxyRules() {
 				"-m", "comment", "--comment", svcName.String(),
 			}
 			// Handle traffic that loops back to the originator with SNAT.
-			// Technically we only need to do this if the endpoint is on this
-			// host, but we don't have that information, so we just do this for
-			// all endpoints.
-			// TODO: if we grow logic to get this node's pod CIDR, we can use it.
-			writeLine(natRules, append(args,
-				"-s", fmt.Sprintf("%s/32", strings.Split(endpoints[i], ":")[0]),
-				"-j", string(KubeMarkMasqChain))...)
-
+			// We only need to do this if the endpoint is on this node.
+			if endpoints[i].localEndpoint {
+				writeLine(natRules, append(args,
+					"-s", fmt.Sprintf("%s/32", strings.Split(endpoints[i].ip, ":")[0]),
+					"-j", string(KubeMarkMasqChain))...)
+			} else {
+				glog.V(3).Infof("Skipping SNAT hairpin rule for non-local endpoint %s for service %s",
+					endpoints[i].ip, svcName.String())
+			}
+			args = []string{
+				"-A", string(endpointChain),
+				"-m", "comment", "--comment",
+				fmt.Sprintf(`"Endpoint rule for %s"`, svcName.String()),
+			}
 			// Update client-affinity lists.
 			if svcInfo.sessionAffinityType == api.ServiceAffinityClientIP {
 				args = append(args, "-m", "recent", "--name", string(endpointChain), "--set")
 			}
 			// DNAT to final destination.
-			args = append(args, "-m", protocol, "-p", protocol, "-j", "DNAT", "--to-destination", endpoints[i])
+			args = append(args, "-m", protocol, "-p", protocol, "-j", "DNAT", "--to-destination", endpoints[i].ip)
+			writeLine(natRules, args...)
+		}
+
+		if !svcInfo.onlyNodeLocalEndpoints {
+			continue
+		}
+		// Now write ingress loadbalancing & DNAT rules only for ESIPP preservation services.
+		// If the service did not have a local-service annotation, this svcLbChain is not needed.
+		localEndpoints := make([]*endpointsInfo, 0)
+		localEndpointChains := make([]utiliptables.Chain, 0)
+		for i, endpointChain := range endpointChains {
+			if endpoints[i].localEndpoint {
+				localEndpoints = append(localEndpoints, endpoints[i])
+				localEndpointChains = append(localEndpointChains, endpointChain)
+			}
+		}
+
+		n = len(localEndpointChains)
+		if n == 0 {
+			// Blackhole all LB traffic that should not be double-hopped.
+			// TODO - nat table does not allow DROP action - use mark action from FW chain (Minhan's commit)
+			args := []string{
+				"-A", string(svcLbChain),
+				"-m", "comment", "--comment",
+				fmt.Sprintf(`"Blackhole rule for %s - No local endpoints"`, svcName.String()),
+				"-j",
+				"LOG",
+			}
 			writeLine(natRules, args...)
+		} else {
+			// Setup probability filter rules only over local endpoints
+			for i, endpointChain := range localEndpointChains {
```

Comment:
Think about: can we merge this with the identical block above?  Maybe a function?